Initial Documentation (#680)
nucypher/nucypher: docs/source/architecture/contracts.md

# NuCypher Ethereum Contracts

## Contract Listing

* `NuCypherToken` ERC20 token contract with additional function - burn own tokens (only for owners)

> * `MinersEscrow` Holds Ursula's stake, stores information about Ursula's activity and assigns a reward for participating in the NuCypher network. The `Issuer` contract is part of the `MinersEscrow` and uses only to split code

Author: hmmm... @szotov?

	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner and the periods during which tokens will be locked for each miner
8. `UserEscrowLibraryLinker`, `UserEscrowProxy` TBD
9. Pre-deposit tokens to the `UserEscrow` if necessary:
	* Create new instance of the `UserEscrow` contract 
	* Transfer ownership of the instance of the `UserEscrow` contract to the user
	* Approve the transfer of tokens for the `UserEscrow`
	* Deposit tokens by the `initialDeposit(uint256, uint256)` method

## Alice's Contract Interaction

### Alice Authors a Blockchain Policy

Alice uses a network of Ursula miners to deploy policies.
In order to take advantage of the network, Alice chooses miners and deploys policies with fees for those miners.
Alice can choose miners by herself ("handpicked") or by using `findCumSum(uint256, uint256, uint256)` method of the contract `MinersEscrow` ("sampling").
`findCumSum` parameters are:

> 	* The start index (if the method is not called the first time)

Reviewer: ```suggestion
* The start index (if the method is not called the first time)
```

> 	* The delta of the step

Reviewer: ```suggestion
* The delta of the step
```

> 	* minimum number of periods during which tokens are locked.

Reviewer: ```suggestion
* minimum number of periods during which tokens are locked.
```

### Alice Revokes a Blockchain Policy

When Alice wants to revoke a policy, she calls the `PolicyManager.revokePolicy(bytes20)` or `PolicyManager.revokeArrangement(bytes20, address)`.
Execution of these methods results in Alice recovering all fees for future periods, and also for periods when the miners were inactive.
Alice can refund ETH for any inactive miners' periods without revoking the policy by using methods `PolicyManager.refund(bytes20)` or `PolicyManager.refund(bytes20, address)`.

## Ursula's Contract Interaction

### Ursula Locks Tokens

In order to become a participant of the network, a miner stakes tokens in the `MinersEscrow` contract.
The miner allows the (mining) contract to perform a transaction using the `MinersEscrow.approve(address, uint256)` method
(ERC20 contracts allow access delegation to another address).

After that, the miner transfers some quantity of tokens (`MinersEscrow.deposit(uint256, uint256)`), locking them at the same time.
Alternately the `NucypherToken.approveAndCall(address, uint256, bytes)` method can be used.
The parameters are:

> 	* The address of the `MinersEscrow` contract

Reviewer: These lines don't show as bullets (whitespace issue)
```suggestion
* The address of the `MinersEscrow` contract
```

> 	* The amount of staked tokens

Reviewer: ```suggestion
* The amount of staked tokens
```

> 	* The periods for locking (which are serialized into an array of bytes)

Reviewer: ```suggestion
* The periods for locking (which are serialized into an array of bytes)
```

The method `MinersEscrow.confirmActivity` is called every time the methods `MinersEscrow.deposit(uint256, uint256)` or `MinersEscrow.lock(uint256, uint256)` is called.
The miner gets a reward for every confirmed period.

### Ursula Generates Staking Rewards
After the period of activity has passed, the miner may call `MinersEscrow.mint()` method which computes and transfers tokens to the miner's account.
Also note that calls to `MinersEscrow.lock(uint256, uint256)` and `MinersEscrow.confirmActivity()` are included the `MinersEscrow.mint()` method.

The reward value depends on the fraction of locked tokens for the period (only those who confirmed activity are accounted for).
Also, the reward depends on the number of periods during which the tokens will be locked: if the tokens will be locked for half a year, the coefficient is 1.5.
The minimum coefficient is 1 (when tokens will get unlocked in the next period), and the maximum is 2 (when the time is 1 year or more).
The reward is calculated separately for each stake that is active during the mining period and all rewards are summed up.
The order of calling `mint` by miners (e.g. who calls first, second etc) doesn't matter.

> Miners can claim their rewards by using the `witdraw(uint256)` method. Only non-locked tokens can be withdrawn.

Reviewer: Did you mean `withdraw`?

> Miners can claim their rewards by using the `witdraw(uint256)` method. Only non-locked tokens can be withdrawn.

Reviewer: Did you mean `withdraw` - missed the `h`

> Miners can claim their rewards by using the `witdraw(uint256)` method. Only non-locked tokens can be withdrawn.

Reviewer 2: ```suggestion
Miners can claim their rewards by using the `withdraw(uint256)` method. Only non-locked tokens can be withdrawn.
```

### Ursula Generates Policy Rewards
Also the miner gets rewards for policies deployed.
Computation of the reward happens every time `mint()` is called by the `updateReward(address, uint256)` method.

> In order to take the reward, the miner needs to call method `withdraw()` of the contract `PolicyManager`.

Reviewer: Does the `withdraw` function here take a parameter?

Reviewer 3: This `withdraw` function without parameters

> In order to take the reward, the miner needs to call method `withdraw()` of the contract `PolicyManager`.

Reviewer: Does `withdraw` take a parameter?

The miner can set a minimum reward rate for a policy. For that, the miner should call the `setMinRewardRate(uint256)` method.

### NuCypher Partner Ursula Staking
Some users will have locked but not staked tokens.
In that case, an instance of the `UserEscrow` contract will hold their tokens (method `initialDeposit(uint256, uint256)`).

> All tokens will be unlocked after a specified time and the user can get them by method `withdraw(uint256)`.

Reviewer: ```suggestion
All tokens will be unlocked after a specified time and the user can retrieve them using the `withdraw(uint256)` method.
```

> All tokens will be unlocked after a specified time and the user can get them by method `withdraw(uint256)`.

Reviewer: ```suggestion
All tokens will be unlocked after a specified time and the user can retrieve them using the `withdraw(uint256)` method.
```
